docs/CppInteroperability/Roadmap.md

```diff
@@ -0,0 +1,141 @@
+
+C++ interoperability aims to provide a clear, well-defined mapping for APIs that can be imported, and to draw a clear boundary between these APIs and APIs that cannot be imported. A user should be able to read the C++ interoperability documentation and have a good idea of both how much of their C++ API will be able to be imported in Swift and what that API will look like once it is imported. 
+
+It is a goal not to have hidden performance traps. C++ APIs should have similar performance in Swift as C++ and should be called directly when possible. Imported APIs, especially containers, should not be implicitly converted to Swift types (in contrast to the automatic Objective-C bridging for NSArray and NSString). 
+
+It is a goal to preserve the semantics (including performance semantics) that an API was written with by understanding the API’s semantics and mapping it to a Swift idiom that represents the same thing. 
+
+Swift programmers have used C++ APIs since Swift 1.0 through an Objective-C or C bridging layer. It is important that programmers are able to *incrementally* remove these (potentially huge) bridging layers and start using their C++ APIs directly. To make this transition incremental, C++ interoperability must not change the way Swift imports any existing C or Objective-C APIs, even when C++ interoperability is enabled. 
+
+It is *not* a goal to import every C++ API into Swift. APIs that are imported by Swift should feel Swifty and idiomatic. There must be an easy way to add annotations so that APIs become importable, especially when updating APIs directly is not an option (because the source code is not available). 
+
+## The approach
+
+The approach for importing C++ APIs into Swift is derived from the fact that both our goals, and C++ as a language, necessitate that not all APIs are imported, and to import an API the user must provide semantic information in some cases. Further, in order to realize our goals (especially the goals of safety and clear, well defined mappings), it is impossible to import any arbitrary C++ API. APIs that are imported by Swift must fall into one of our understood categories or be annotated with semantic information that can be used to classify them into a Swift idiom. 
+
+In this section, the roadmap will discuss several of the most common C++ API patterns and how they might be imported by the Swift compiler using the goals outlined above.
+
+### Definitions
+
+**Reference types** have reference semantics and object identity. A reference type is a pointer (or “reference”) to some object which means there is a layer of indirection. When a reference type is copied, the pointer’s value is copied rather than the object’s storage. This means reference types can be used to represent non-copyable types in C++.
```

Comment:
You've set me up with the introduction that we're about to “discuss several of the most common C++ API patterns,” so when I get here I understand you to be categorizing C++ types.  But the last sentence above leads to the opposite understanding.

Reply:
> reference types can be used to represent non-copyable types in C++

This idea conflicts with the idea that things should not become less safe when imported into Swift.  In C++, non-copyable types are idiomatically used to enforce single-ownership semantics, on which the safety of other constructs may depend.  Representing them directly as copyable references in Swift undermines that.  One solution could be to import a non-copyable C++ `T` as a class `Shared<T>` in Swift, but the API necessary to make the indirection evident will be ugly until Swift itself gets new features… like at least partial direct support for non-copyable types.